File: runner/src/integration-tests/java/com/daylifecraft/minigames/chat/InstanceChatEventTest.java
package com.daylifecraft.minigames.chat;

import com.daylifecraft.minigames.ChatManager;
import com.daylifecraft.minigames.Init;
import com.daylifecraft.minigames.UtilsForTesting;
import net.kyori.adventure.text.Component;
import net.minestom.server.entity.Player;
import net.minestom.server.event.EventDispatcher;
import net.minestom.server.event.player.PlayerChatEvent;
import org.junit.jupiter.api.*;
import org.mockito.ArgumentMatchers;
import org.mockito.Mockito;

import java.util.Collections;

@TestMethodOrder(MethodOrderer.OrderAnnotation.class)
class InstanceChatEventTest {

  private static Player fakePlayer;

  private static final ChatManager chatManager = Mockito.mock(ChatManager.class);

  @BeforeAll
  static void start() throws InterruptedException {
    Init.setupChatManager(chatManager);

    fakePlayer = UtilsForTesting.initFakePlayer("EventsTest");

    UtilsForTesting.waitUntilPlayerJoin(fakePlayer);
  }

  @Test
  @Order(1)
  void testDoesEventCalled() {
    final PlayerChatEvent event =
      new PlayerChatEvent(fakePlayer, Collections.emptyList(), Component::empty, "Some text");
    EventDispatcher.call(event);

    Assertions.assertTrue(event.isCancelled(), "Event must me cancelled");

    Mockito.verify(chatManager)
      .sendPlayerChatMessage(
        ArgumentMatchers.eq(fakePlayer),
        ArgumentMatchers.eq(fakePlayer),
        ArgumentMatchers.anyString());
  }

  @AfterAll
  static void kickPlayer() {
    fakePlayer.kick("");
  }
}

File: runner/src/integration-tests/java/com/daylifecraft/minigames/chat/InstanceChatEventTest.kt
package com.daylifecraft.minigames.chat

import com.daylifecraft.minigames.ChatManager
import com.daylifecraft.minigames.Init.setupChatManager
import com.daylifecraft.minigames.UtilsForTesting
import io.mockk.mockk
import io.mockk.verify
import net.kyori.adventure.text.Component
import net.minestom.server.entity.Player
import net.minestom.server.event.EventDispatcher
import net.minestom.server.event.player.PlayerChatEvent
import org.junit.jupiter.api.AfterAll
import org.junit.jupiter.api.BeforeAll
import org.junit.jupiter.api.Test
import kotlin.test.assertTrue

internal class InstanceChatEventTest {
  @Test
  fun testDoesEventCalled() {
    val event = PlayerChatEvent(fakePlayer, emptyList(), { Component.empty() }, "Some text")
    EventDispatcher.call(event)

    assertTrue(event.isCancelled, "Event must me cancelled")

    verify {
      chatManager.sendPlayerChatMessage(
        fakePlayer,
        fakePlayer,
        any(),
      )
    }
  }

  companion object {
    private lateinit var fakePlayer: Player

    private val chatManager = mockk<ChatManager>(relaxed = true)

    @BeforeAll
    @JvmStatic
    fun start() {
      setupChatManager(chatManager)

      fakePlayer = UtilsForTesting.initFakePlayer("EventsTest")

      UtilsForTesting.waitUntilPlayerJoin(fakePlayer)
    }

    @AfterAll
    @JvmStatic
    fun kickPlayer() {
      fakePlayer.kick("")
    }
  }
}
